document/js-api/index.bs

```diff
@@ -1120,26 +1144,155 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
 1. If |type| is [=f64=],
     1. Let |f64| be ? [=ToNumber=](|v|).
     1. Return [=f64.const=] |f64|.
-1. If |type| is [=funcref=],
+1. If |type| is of the form [=ref=] |null| |heaptype| and is a subtype of [=ref=] |null| [=heap-type/func=],
     1. If |v| is null,
-        1. Return [=ref.null=] [=funcref=].
+        1. If |null| is present,
+            1. Return [=ref.null=] |heaptype|.
+        1. Otherwise,
+            1. Throw a {{TypeError}}.
     1. If |v| is an [=Exported Function=],
         1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
-        1. Return [=ref.func=] |funcaddr|.
+        1. Let |casted| be [=ref_cast=](|heaptype|, |funcaddr|).
```

Comment:
Casts do not change the value, so if the cast succeeds, it would simply return the original funcaddr. But the cast may also fail, so the result type of this operator would have to include an error case.

Hence it is probably more natural for the embedding API to provide a predicate like
```
ref_typecheck(addr, heaptype) : bool
```
instead. (Perhaps we even want to separate getting the ref's own type from a subtype function on two types.)

Reply:
Thanks for catching that, I wonder if it would make sense to call the operation `ref_test` to mirror the `ref.test` instruction that is similar in functionality.

Reply:
The latest commit uses `ref_test` for now (not tied to this naming, could use something else or go with a different way of checking) and should check the failure case.